### PR TITLE
Add config_dir option to config objects

### PR DESCRIPTION
**Pull request by shannynalayna**
Adds a `config_dir` option to the DefaultPartialConfigBuilder, EnvPartialConfigBuilder, and the ClapPartialConfigBuilder. The `config_dir`, default `/etc/splinter`, is appended to the beginning of the `tls_cert_dir` if the `tls_cert_dir` does not begin with "/" (in which case we assume this is the full path)

This pr also adds another step to building the config file paths to check if the file starts with "/", in which case we assume this is the full path. If it does not start with "/", the default / set tls cert directory will be appended to the beginning of the file. 

This also updates the tests to reflect these changes.

### CODE REVIEW

File: splinterd/src/config/mod.rs [hunk 1]
```diff
@@ -425,6 +425,24 @@ impl Config {
...
+            self.tls_cert_dir(),
+            self.tls_cert_dir_source()
+        );
+        debug!(
+            "Config: state_dir: {} (source: {:?})",
+            self.state_dir(),
+            self.state_dir_source()
+        );
+    }
 }
```
Comment (ltseeley) [inline splinterd/src/config/mod.rs:446]: Seems unnecessary to split this out from the main function

Comment (shannynalayna) [reply splinterd/src/config/mod.rs:446]: There was a clippy error when I added the whitelist debug statement about the complexity of the function being too high. This was an attempt to not run into that clippy error again.

Comment (ltseeley) [reply splinterd/src/config/mod.rs:446]: Hmm, this might actually be a case where ignoring the lint is the right thing to do. The cognitive complexity of the `log_as_debug` function is actually very low since it only does one thing; it just does it many times. IMO splitting parts of it out just makes it more complex, because you have to look at the various sub-functions to get the full picture.

Comment (shannynalayna) [reply splinterd/src/config/mod.rs:446]: Mmkay, ya I would agree. I'll fix that

File: splinterd/src/config/env.rs [hunk 2]
```diff
@@ -16,6 +16,7 @@ use std::env;
 
 use crate::config::{ConfigError, ConfigSource, PartialConfig, PartialConfigBuilder};
 
+const CONFIG_DIR: &str = "CONFIG_DIR";
```
Comment (vaporos) [inline splinterd/src/config/env.rs:19]: Following the convention of the other two environment variables - the name of the const should be CONFIG_DIR_ENV and the environment variable should be SPLINTER_CONFIG_DIR.

File: splinterd/src/config/default.rs [hunk 3]
```diff
@@ -14,8 +14,9 @@
 
 use crate::config::{ConfigError, ConfigSource, PartialConfig, PartialConfigBuilder};
 
-const DEFAULT_CERT_DIR: &str = "/etc/splinter/certs/";
-const DEFAULT_STATE_DIR: &str = "/var/lib/splinter/";
+pub(in crate::config) const CONFIG_DIR: &str = "/etc/splinter";
```
Comment (vaporos) [inline splinterd/src/config/default.rs:17]: These should remain private to this module.

Comment (shannynalayna) [reply splinterd/src/config/default.rs]: Fixed

File: splinterd/src/config/builder.rs [hunk 4]
```diff
@@ -23,15 +23,23 @@ pub trait PartialConfigBuilder {
     fn build(self) -> Result<PartialConfig, ConfigError>;
 }
 
-fn get_file_path(cert_dir: &str, file: &str) -> String {
+fn get_tls_file_path(config_dir: &str, cert_dir: &str, file: &str) -> String {
     if file.starts_with("./") || file.starts_with("../") {
```
Comment (vaporos) [inline splinterd/src/config/builder.rs]: Should this check for startswith('/') too? I think yes, in that if the user specified the full path "/tmp/foobar" then we don't want to prefix it with anything.

Comment (shannynalayna) [reply splinterd/src/config/builder.rs]: That is addressed in the second commit

File: splinterd/man/splinterd.1.md [hunk 5]
```diff
@@ -112,6 +112,13 @@ OPTIONS
   Any options on the command line will override the settings in the
   configuration file.
 
+`--config-dir CONFIG_DIR`
+: Specifies the directory containing configuration files. This is appended to
+  the directory containing certificates and associated key files if an absolute
```
Comment (vaporos) [inline splinterd/man/splinterd.1.md:117]: Technically "absolute path" is one that starts with "/", so this should be stated in a different way.

File: splinterd/src/config/builder.rs [hunk 6]
```diff
@@ -23,15 +23,23 @@ pub trait PartialConfigBuilder {
...
         String::from(file)
-    } else {
+    } else if cert_dir.starts_with('/') || cert_dir.starts_with("./") || cert_dir.starts_with("../")
```
Comment (vaporos) [inline splinterd/src/config/builder.rs:29]: A couple thoughts:

1. The absolute path check should be done with std::path::Path's absolute() check so it is cross-platform compatible.
2. The check for "./" and "../" should also check for the exact strings "." and ".." which would also be relative paths that we do not want to prefix. Really, then, we should split the path and (using stuff in std::path) and do this comparison on the first element of the path instead of start_with().

File: splinterd/src/config/default.rs [hunk 7]
```diff
@@ -14,8 +14,9 @@
 
 use crate::config::{ConfigError, ConfigSource, PartialConfig, PartialConfigBuilder};
 
-const DEFAULT_CERT_DIR: &str = "/etc/splinter/certs/";
-const DEFAULT_STATE_DIR: &str = "/var/lib/splinter/";
+pub const CONFIG_DIR: &str = "/etc/splinter";
```
Comment (vaporos) [inline splinterd/src/config/default.rs:17]: This hsould be DEFAULT_CONFIG_DIR.

The pub should be removed here and on the following two lines.

File: splinterd/src/config/mod.rs [hunk 8]
```diff
@@ -287,8 +296,24 @@ impl Config {
...
 
+    #[allow(clippy::cognitive_complexity)]
     /// Displays the configuration value along with where the value was sourced from.
     pub fn log_as_debug(&self) {
+        debug!(
+            "Config: config_dir: {:?} (source: {:?})",
+            self.config_dir(),
+            self.config_dir_source()
+        );
+        debug!(
```
Comment (vaporos) [inline splinterd/src/config/mod.rs:307]: These were previously sorted alphabetically. That is probably still the best approach here?

File: splinterd/src/config/mod.rs [hunk 9]
```diff
@@ -412,17 +427,14 @@ impl Config {
...
-
-    #[cfg(feature = "rest-api-cors")]
-    fn log_whitelist(&self) {
-        if let Some(list) = self.whitelist() {
-            debug!(
-                "Config: whitelist: {:?} (source: {:?})",
-                list,
-                self.whitelist_source()
-            );
+        {
```
Comment (vaporos) [inline splinterd/src/config/mod.rs:430]: This change looks unrelated.

File: splinterd/src/config/mod.rs (hunk 8 above)
Comment (shannynalayna) [reply splinterd/src/config/mod.rs:306]: I don't think these were previously sorted alphabetically, it seems like they've always been sorted in the order the entry is in the Config object. I've added the log for the config_dir first, as this is now the first entry in the Config objects.

File: splinterd/man/splinterd.1.md [hunk 10]
```diff
@@ -112,6 +112,10 @@ OPTIONS
   Any options on the command line will override the settings in the
   configuration file.
 
+`--config-dir CONFIG_DIR`
+: Specifies the directory containing configuration files.
+  (Default: `/etc/splinter`.)
```
Comment (chenette) [inline splinterd/man/splinterd.1.md:117]: Make consistent with other options:
- `CONFIG-DIR` (with a hyphen)
- Mention the `SPLINTER_CONFIG_DIR` environment variable here, following the model of `--tls-ca-file` (precede with a blank line and indent so it lines up with the existing description).

Comment (shannynalayna) [reply splinterd/man/splinterd.1.md:117]: Updated.
